subprojects/file-watching/src/main/java/org/gradle/internal/watch/vfs/impl/WatchingVirtualFileSystem.java

```diff
@@ -117,8 +119,36 @@ public void beforeBuildFinished(boolean watchingEnabled) {
         if (watchingEnabled) {
             getRoot().update(currentRoot -> {
                 buildRunning = false;
+                Set<String> symlinkLocations = new HashSet<>();
```

Comment:
Why collect these in a `Set` first and then invalidate instead of invalidating directly?

Also, if we need to collect them, collecting `CompleteFileSystemLocationSnapshot`s and resolving the absolute path later sounds better to me. That way we keep track of higher level things instead of just `String`s.

Reply:
Some comments on why we are doing what we are doing here would be appreciated, too.

Alternatively, I wonder if we could capture the symlink-related functionality in a separate decorator class or something. That way we could use Javadoc on that class to document the process (and then replace it later on with canonical symlink traking maybe).

Reply:
I am now invalidating the locations directly. I also added some comments in https://github.com/gradle/gradle/issues/13053.